connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/GrpcRetryHandler.scala

```diff
@@ -148,37 +151,58 @@ private[client] object GrpcRetryHandler extends Logging {
 
   /**
    * Retries the given function with exponential backoff according to the client's retryPolicy.
+   *
    * @param retryPolicy
    *   The retry policy
+   * @param sleep
+   *   The function which sleeps (takes number of milliseconds to sleep)
    * @param fn
    *   The function to retry.
-   * @param currentRetryNum
-   *   Current number of retries.
    * @tparam T
    *   The return type of the function.
    * @return
    *   The result of the function.
    */
-  @tailrec final def retry[T](retryPolicy: RetryPolicy)(fn: => T, currentRetryNum: Int = 0): T = {
-    if (currentRetryNum > retryPolicy.maxRetries) {
-      throw new IllegalArgumentException(
-        s"The number of retries ($currentRetryNum) must not exceed " +
-          s"the maximum number of retires (${retryPolicy.maxRetries}).")
-    }
-    try {
-      return fn
-    } catch {
-      case NonFatal(e)
-          if (retryPolicy.canRetry(e) || e.isInstanceOf[RetryException])
-            && currentRetryNum < retryPolicy.maxRetries =>
-        logWarning(
-          s"Non fatal error during RPC execution: $e, " +
-            s"retrying (currentRetryNum=$currentRetryNum)")
-        Thread.sleep(
-          (retryPolicy.maxBackoff min retryPolicy.initialBackoff * Math
-            .pow(retryPolicy.backoffMultiplier, currentRetryNum)).toMillis)
+  final def retry[T](retryPolicy: RetryPolicy, sleep: Long => Unit = Thread.sleep)(
+      fn: => T): T = {
+    var currentRetryNum = 0
+    var lastException: Throwable = null
+    var nextBackoff: Duration = retryPolicy.initialBackoff
+
+    while (currentRetryNum <= retryPolicy.maxRetries) {
+      if (currentRetryNum != 0) {
+        var currentBackoff = nextBackoff
+        if (currentBackoff >= retryPolicy.minJitterThreshold) {
+          currentBackoff += Random.nextDouble() * retryPolicy.jitter
+        }
+        nextBackoff = nextBackoff * retryPolicy.backoffMultiplier min retryPolicy.maxBackoff
+
+        sleep(currentBackoff.toMillis)
+      }
+
+      try {
+        return fn
+      } catch {
+        case NonFatal(e) if retryPolicy.canRetry(e) && currentRetryNum < retryPolicy.maxRetries =>
+          currentRetryNum += 1
+          lastException = e
+
+          if (currentRetryNum <= retryPolicy.maxRetries) {
+            logWarning(
+              s"Non-Fatal error during RPC execution: $e, " +
+                s"retrying (currentRetryNum=$currentRetryNum)")
+          } else {
+            logWarning(
+              s"Non-Fatal error during RPC execution: $e, " +
+                s"exceeded retries (currentRetryNum=$currentRetryNum)")
+          }
+      }
     }
-    retry(retryPolicy)(fn, currentRetryNum + 1)
+
+    throw new SparkException(
+      errorClass = "EXCEED_RETRY",
```

Comment:
Class "EXCEED_RETRY" does not exist in error-classes.json, so this will fail in SparkThrowableHelper.getMessage and turn it into an internal error "Undefined error message parameter for error class".

Reply:
Class "EXCEED_RETRY" does exist in `python/pyspark/errors/error_classes.py` for the python side... but on the other hand various classes added with query cancellation and reattachable execution (INVALID_HANDLE, OPERATION_CANCELED) do not exist there? @HyukjinKwon do these errors added to error-classes.json need to be duplicated into error_classes.py in Pyspark?

Reply:
Yeah it's duplicated for now iirc. Cc @itholic to confirm

Reply:
Thanks.
This PR should add EXCEED_RETRY to error-classes.json then, and then see `SparkThrowableSuite` to auto-regenerate error documentation from it.
A followup PR should add INVALID_HANDLE and OPERATION_CANCELED to python. @itholic @HyukjinKwon since SparkThrowableSuite already has some automation to validate that error-classes.json is in sync with documentation, would it be a good idea to add such validation, and auto-generation for error_classes.py?

Reply:
> A followup PR should add INVALID_HANDLE and OPERATION_CANCELED to python.

Yup.

> since SparkThrowableSuite already has some automation to validate that error-classes.json is in sync with documentation, would it be a good idea to add such validation, and auto-generation for error_classes.py?

Yes, I think we should.

Reply:
Yes, I think we should have the validation and checking there.

Reply:
Actually we already have a [separate error class documentation for PySpark](https://github.com/apache/spark/blob/master/python/docs/source/development/errors.rst), which will be published from Spark 3.5.

I believe we can't simply duplicate the error classes from PySpark to the JVM, because basically the error class on the SQL side is an SQL-related error that can assign an SQLSTATE, whereas the error class on the PySpark side is mainly Python type and value related errors that are not related to SQL.

For example, error class from PySpark such as `CANNOT_ACCESS_TO_DUNDER` or `SLICE_WITH_STEP` would never be used from JVM, so it's not make sense to sync all the error classes from both side.

**So, the solution I suggest is that:**
- Add a test on the PySpark side to check if the name of the error class does not overlap with the JVM side.
- Rename the duplicated error classes. In this case, we can use `EXCEED_RETRY_JVM` or something like that to avoid duplication from PySpark side.

Reply:
@itholic @HyukjinKwon Are you sure they can't have the same name? Maybe they are different domain that can coexist with the same name for two different things? It's kind of ugly to have this _JVM in the user facing name?

Reply:
On my second thought, I think it's okay to have the same error class name if the error message of the error class is the same.

Reply:
@itholic we should set the policy here, document, and guide other contributors to follow it. Either duplicating all between Python and JVM, or making them mutually exclusive works to me once we set it down.

Reply:
In any case, I find `EXCEED_RETRY_JVM` ugly for a user facing name, and if it needs to be different, I'd rather have some "workaround". There is precedent like e.g. UNSUPPORTED_DATA_TYPE in python vs UNSUPPORTED_DATATYPE in scala...

Reply:
Just had a brief discussion on offline, it seems that it doesn't matter even if the error class name is duplicated because it can be distinguished by the package name since we basically manage JVM and Python errors in different ways. Sorry for the confusion 🙏 

So, we can just keep using `EXCEED_RETRY` if you believe that it is the name of the error class `EXCEED_RETRY` that best expresses the currently occurring error (however, I think the error message still needs to be modified. Using the Python error message as it is does not seem appropriate for this situation)

Reply:
@cdkrot I looked again at the python code
https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/client/core.py#L1638C1-L1647C22
```
            # If the number of retries have exceeded the maximum allowed retries.
            if retry_state.count() > self._max_retries:
                e = retry_state.exception()
                if e is not None:
                    raise e
                else:
                    raise PySparkRuntimeError(
                        error_class="EXCEED_RETRY",
                        message_parameters={},
                    )
```
it seems that if there was an exception, it just rethrows the exception. It throws the EXCEED_RETRY exception only if the number of retries has been exceeded, but there is no exception saved in the retry state (`retry_state.exception()`).  I don't think that situation is even possible, because in https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/client/core.py#L1578 we retry only ever if there is an exception. So the situation throwing this EXCEED_RETRY error in python is only occuring if the code is faulty with invalid use of RetryState.set_exception, and it is an internal error. And then the message "Retries exceeded but no exception caught." actually fits the situation.
@HyukjinKwon @itholic should this be its error class in python if it's an internal should-never-happen assertion?

So I think this error class isn't needed in scala (there is always an underlying exception available to throw at this place in the code), and we should just rethrow the original exception like python does.

Reply:
Agree, that it doesn't seem possible. Will address